File: EventOrganizer.java
// EventOrganizer.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import com.google.gson.*;
import com.google.zxing.BarcodeFormat;
import com.google.zxing.client.j2se.MatrixToImageWriter;
import com.google.zxing.common.BitMatrix;
import com.google.zxing.qrcode.QRCodeWriter;

class Attendee {
    String id, name, email;
    boolean checked_in;
}

class Event {
    String id, name, date, venue;
    List<Attendee> attendees = new ArrayList<>();
}

public class EventOrganizer {
    private List<Event> events = new ArrayList<>();
    private final String dataFile = "events.json";
    private final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    public EventOrganizer() { load(); }

    private void load() {
        try {
            Path path = Paths.get(dataFile);
            if (Files.exists(path)) {
                String json = new String(Files.readAllBytes(path));
                Event[] arr = gson.fromJson(json, Event[].class);
                events = Arrays.asList(arr);
            }
        } catch (Exception e) {}
    }

    private void save() {
        try {
            Files.write(Paths.get(dataFile), gson.toJson(events).getBytes());
        } catch (Exception e) {}
    }

    private String generateId() { return UUID.randomUUID().toString().substring(0,8); }

    public void createEvent(String name, String date, String venue) {
        Event e = new Event();
        e.id = generateId();
        e.name = name;
        e.date = date;
        e.venue = venue;
        events.add(e);
        save();
        System.out.printf("Event created: %s - %s%n", e.id, e.name);
    }

    public void listEvents() {
        if (events.isEmpty()) { System.out.println("No events."); return; }
        System.out.println("\n📅 Events:");
        for (Event e : events) {
            System.out.printf("  %s: %s (%s) at %s - %d attendees%n", e.id, e.name, e.date, e.venue, e.attendees.size());
        }
    }

    public void addAttendee(String eventId, String name, String email) {
        for (Event e : events) {
            if (e.id.equals(eventId)) {
                Attendee a = new Attendee();
                a.id = generateId();
                a.name = name;
                a.email = email;
                a.checked_in = false;
                e.attendees.add(a);
                save();
                System.out.printf("Attendee %s added with ticket ID %s%n", name, a.id);
                return;
            }
        }
        System.out.printf("Event %s not found.%n", eventId);
    }

    public void listAttendees(String eventId) {
        for (Event e : events) {
            if (e.id.equals(eventId)) {
                if (e.attendees.isEmpty()) { System.out.println("No attendees."); return; }
                System.out.printf("\n👤 Attendees for %s:%n", e.name);
                for (Attendee a : e.attendees) {
                    String status = a.checked_in ? "✓" : "✗";
                    System.out.printf("  %s: %s (%s) - %s%n", a.id, a.name, a.email, status);
                }
                return;
            }
        }
        System.out.printf("Event %s not found.%n", eventId);
    }

    public void generateTickets(String eventId, String outputDir) {
        for (Event e : events) {
            if (e.id.equals(eventId)) {
                try {
                    Files.createDirectories(Paths.get(outputDir));
                    QRCodeWriter qrWriter = new QRCodeWriter();
                    for (Attendee a : e.attendees) {
                        String data = e.id + ":" + a.id;
                        BitMatrix matrix = qrWriter.encode(data, BarcodeFormat.QR_CODE, 300, 300);
                        Path path = Paths.get(outputDir, a.id + ".png");
                        MatrixToImageWriter.writeToPath(matrix, "PNG", path);
                        System.out.printf("Ticket saved for %s: %s%n", a.name, path.toString());
                    }
                } catch (Exception ex) {
                    System.err.println("Error generating tickets: " + ex.getMessage());
                }
                return;
            }
        }
        System.out.printf("Event %s not found.%n", eventId);
    }

    public void checkin(String eventId, String ticketId) {
        for (Event e : events) {
            if (e.id.equals(eventId)) {
                for (Attendee a : e.attendees) {
                    if (a.id.equals(ticketId)) {
                        if (a.checked_in) {
                            System.out.printf("Attendee %s already checked in.%n", a.name);
                        } else {
                            a.checked_in = true;
                            save();
                            System.out.printf("Checked in %s.%n", a.name);
                        }
                        return;
                    }
                }
                System.out.printf("Ticket %s not found for this event.%n", ticketId);
                return;
            }
        }
        System.out.printf("Event %s not found.%n", eventId);
    }

    public void status(String eventId) {
        for (Event e : events) {
            if (e.id.equals(eventId)) {
                int total = e.attendees.size();
                int checked = 0;
                for (Attendee a : e.attendees) if (a.checked_in) checked++;
                System.out.printf("\n📊 Event: %s%n", e.name);
                System.out.printf("  Total attendees: %d%n", total);
                System.out.printf("  Checked in: %d%n", checked);
                System.out.printf("  Remaining: %d%n", total - checked);
                listAttendees(eventId);
                return;
            }
        }
        System.out.printf("Event %s not found.%n", eventId);
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: EventOrganizer <command> [options]");
            return;
        }
        EventOrganizer app = new EventOrganizer();
        String cmd = args[0];
        switch (cmd) {
            case "event":
                if (args.length < 2) { System.out.println("event list | event create <name> <date> <venue>"); return; }
                if (args[1].equals("list")) app.listEvents();
                else if (args[1].equals("create") && args.length == 5)
                    app.createEvent(args[2], args[3], args[4]);
                else System.out.println("Invalid event subcommand.");
                break;
            case "attendee":
                if (args.length < 3) { System.out.println("attendee list <event_id> | attendee add <event_id> <name> <email>"); return; }
                if (args[1].equals("list")) app.listAttendees(args[2]);
                else if (args[1].equals("add") && args.length == 5)
                    app.addAttendee(args[2], args[3], args[4]);
                else System.out.println("Invalid attendee subcommand.");
                break;
            case "ticket":
                if (args.length < 3 || !args[1].equals("generate")) {
                    System.out.println("ticket generate <event_id> [-o output_dir]");
                    return;
                }
                String outputDir = ".";
                for (int i = 2; i < args.length; i++) {
                    if (args[i].equals("-o") && i+1 < args.length) {
                        outputDir = args[i+1];
                        i++;
                    }
                }
                app.generateTickets(args[2], outputDir);
                break;
            case "checkin":
                if (args.length < 3) { System.out.println("checkin <event_id> <ticket_id>"); return; }
                app.checkin(args[1], args[2]);
                break;
            case "status":
                if (args.length < 2) { System.out.println("status <event_id>"); return; }
                app.status(args[1]);
                break;
            default:
                System.out.println("Unknown command.");
        }
    }
}

File: event_organizer.go
// event_organizer.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"strings"
	"time"
	"github.com/google/uuid"
	"github.com/skip2/go-qrcode"
)

type Attendee struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Email     string `json:"email"`
	CheckedIn bool   `json:"checked_in"`
}

type Event struct {
	ID        string     `json:"id"`
	Name      string     `json:"name"`
	Date      string     `json:"date"`
	Venue     string     `json:"venue"`
	Attendees []Attendee `json:"attendees"`
}

type Organizer struct {
	Events []Event `json:"events"`
	File   string
}

func NewOrganizer(file string) *Organizer {
	o := &Organizer{File: file}
	o.load()
	return o
}

func (o *Organizer) load() {
	data, err := os.ReadFile(o.File)
	if err != nil {
		return
	}
	json.Unmarshal(data, o)
}

func (o *Organizer) save() {
	data, _ := json.MarshalIndent(o, "", "  ")
	os.WriteFile(o.File, data, 0644)
}

func (o *Organizer) CreateEvent(name, date, venue string) {
	ev := Event{
		ID:    uuid.New().String()[:8],
		Name:  name,
		Date:  date,
		Venue: venue,
	}
	o.Events = append(o.Events, ev)
	o.save()
	fmt.Printf("Event created: %s - %s\n", ev.ID, ev.Name)
}

func (o *Organizer) ListEvents() {
	if len(o.Events) == 0 {
		fmt.Println("No events.")
		return
	}
	fmt.Println("\n📅 Events:")
	for _, e := range o.Events {
		fmt.Printf("  %s: %s (%s) at %s - %d attendees\n", e.ID, e.Name, e.Date, e.Venue, len(e.Attendees))
	}
}

func (o *Organizer) AddAttendee(eventID, name, email string) {
	for i, ev := range o.Events {
		if ev.ID == eventID {
			a := Attendee{
				ID:        uuid.New().String()[:8],
				Name:      name,
				Email:     email,
				CheckedIn: false,
			}
			o.Events[i].Attendees = append(o.Events[i].Attendees, a)
			o.save()
			fmt.Printf("Attendee %s added with ticket ID %s\n", name, a.ID)
			return
		}
	}
	fmt.Printf("Event %s not found.\n", eventID)
}

func (o *Organizer) ListAttendees(eventID string) {
	for _, ev := range o.Events {
		if ev.ID == eventID {
			if len(ev.Attendees) == 0 {
				fmt.Println("No attendees.")
				return
			}
			fmt.Printf("\n👤 Attendees for %s:\n", ev.Name)
			for _, a := range ev.Attendees {
				status := "✗"
				if a.CheckedIn {
					status = "✓"
				}
				fmt.Printf("  %s: %s (%s) - %s\n", a.ID, a.Name, a.Email, status)
			}
			return
		}
	}
	fmt.Printf("Event %s not found.\n", eventID)
}

func (o *Organizer) GenerateTickets(eventID, outputDir string) {
	for _, ev := range o.Events {
		if ev.ID == eventID {
			if err := os.MkdirAll(outputDir, 0755); err != nil {
				fmt.Println("Error creating output dir:", err)
				return
			}
			for _, a := range ev.Attendees {
				data := fmt.Sprintf("%s:%s", ev.ID, a.ID)
				filename := fmt.Sprintf("%s/%s.png", outputDir, a.ID)
				err := qrcode.WriteFile(data, qrcode.Medium, 256, filename)
				if err != nil {
					fmt.Printf("Error generating ticket for %s: %v\n", a.Name, err)
				} else {
					fmt.Printf("Ticket saved for %s: %s\n", a.Name, filename)
				}
			}
			return
		}
	}
	fmt.Printf("Event %s not found.\n", eventID)
}

func (o *Organizer) Checkin(eventID, ticketID string) {
	for i, ev := range o.Events {
		if ev.ID == eventID {
			for j, a := range ev.Attendees {
				if a.ID == ticketID {
					if a.CheckedIn {
						fmt.Printf("Attendee %s already checked in.\n", a.Name)
					} else {
						o.Events[i].Attendees[j].CheckedIn = true
						o.save()
						fmt.Printf("Checked in %s.\n", a.Name)
					}
					return
				}
			}
			fmt.Printf("Ticket %s not found for this event.\n", ticketID)
			return
		}
	}
	fmt.Printf("Event %s not found.\n", eventID)
}

func (o *Organizer) Status(eventID string) {
	for _, ev := range o.Events {
		if ev.ID == eventID {
			total := len(ev.Attendees)
			checked := 0
			for _, a := range ev.Attendees {
				if a.CheckedIn {
					checked++
				}
			}
			fmt.Printf("\n📊 Event: %s\n", ev.Name)
			fmt.Printf("  Total attendees: %d\n", total)
			fmt.Printf("  Checked in: %d\n", checked)
			fmt.Printf("  Remaining: %d\n", total-checked)
			o.ListAttendees(eventID)
			return
		}
	}
	fmt.Printf("Event %s not found.\n", eventID)
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: event_organizer <command> [options]")
		return
	}
	org := NewOrganizer("events.json")
	switch os.Args[1] {
	case "event":
		if len(os.Args) < 3 {
			fmt.Println("Usage: event create <name> <date> <venue> | event list")
			return
		}
		if os.Args[2] == "list" {
			org.ListEvents()
		} else if os.Args[2] == "create" && len(os.Args) >= 6 {
			org.CreateEvent(os.Args[3], os.Args[4], os.Args[5])
		} else {
			fmt.Println("Invalid event command.")
		}
	case "attendee":
		if len(os.Args) < 4 {
			fmt.Println("Usage: attendee add <event_id> <name> <email> | attendee list <event_id>")
			return
		}
		if os.Args[2] == "list" {
			org.ListAttendees(os.Args[3])
		} else if os.Args[2] == "add" && len(os.Args) >= 6 {
			org.AddAttendee(os.Args[3], os.Args[4], os.Args[5])
		} else {
			fmt.Println("Invalid attendee command.")
		}
	case "ticket":
		if len(os.Args) < 4 {
			fmt.Println("Usage: ticket generate <event_id> [-o output_dir]")
			return
		}
		if os.Args[2] == "generate" {
			outputDir := "."
			for i := 3; i < len(os.Args); i++ {
				if os.Args[i] == "-o" && i+1 < len(os.Args) {
					outputDir = os.Args[i+1]
					i++
				}
			}
			org.GenerateTickets(os.Args[3], outputDir)
		}
	case "checkin":
		if len(os.Args) < 4 {
			fmt.Println("Usage: checkin <event_id> <ticket_id>")
			return
		}
		org.Checkin(os.Args[2], os.Args[3])
	case "status":
		if len(os.Args) < 3 {
			fmt.Println("Usage: status <event_id>")
			return
		}
		org.Status(os.Args[2])
	default:
		fmt.Println("Unknown command.")
	}
}
